custom_random_forest.py

```diff
@@ -0,0 +1,171 @@
+
+
+class RandomForestClassifierCustom(BaseEstimator):
+    """
+    Custom implementation of Random Forest classifier using multiple processes for parallelization.
+
+    Parameters:
+    - n_estimators (int): Number of trees in the forest.
+    - max_depth (int): Maximum depth of the trees.
+    - max_features (int): Maximum number of features considered for splitting a node.
+    - random_state (int): Seed used by the random number generator.
+
+    Attributes:
+    - trees (list): List of trained decision trees.
+    - feat_ids_by_tree (list): List of feature indices used by each tree.
+
+    Methods:
+    - fit(X, y, n_jobs=1): Fit the model to the training data.
+    - _fit_process(n_estimators, X, y, workers_id): Helper function for fitting trees in parallel.
+    - predict_proba(X, n_jobs): Predict class probabilities for the input samples.
+    - _predict_proba_process(n_estimators, X, workers_id): Helper function for predicting probabilities in parallel.
+    - predict(X, n_jobs=1): Predict class labels for the input samples.
+    """
+
+    def __init__(
+        self, n_estimators=10, max_depth=None, max_features=None, random_state=42
+    ):
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.max_features = max_features
+        self.random_state = random_state
+        self.trees = []
+        self.feat_ids_by_tree = []
+
+    def fit(self, X, y, n_jobs=1):
+        """
+        Fit the Random Forest classifier to the training data.
+
+        Parameters:
+        - X (array-like): Training data.
+        - y (array-like): Target values.
+        - n_jobs (int): Number of processes to use for parallelization.
+        """
+        self.classes_ = sorted(np.unique(y))
+        n_estimators_subset_size = self.n_estimators // n_jobs
+        workers_id = list(range(n_jobs))
+        with ProcessPoolExecutor(n_jobs) as pool:
+            results = pool.map(
+                self._fit_process,
+                [n_estimators_subset_size] * n_jobs,
+                [X] * n_jobs,
+                [y] * n_jobs,
+                workers_id,
+            )
+            for result, feat_ids_list in results:
+                self.trees.extend(list(result))
+                self.feat_ids_by_tree.extend(feat_ids_list)
+
+    def _fit_process(self, n_estimators, X, y, workers_id):
+        """
+        Helper function to fit trees in parallel.
+
+        Parameters:
+        - n_estimators (int): Number of trees to fit.
+        - X (array-like): Training data.
+        - y (array-like): Target values.
+        - workers_id (int): Identifier for the current worker process.
+
+        Returns:
+        - trees (list): List of trained decision trees.
+        - feat_ids_list (list): List of feature indices used by each tree.
+        """
+        trees = []
+        feat_ids_list = []
+
+        for i in range(n_estimators):
+            np.random.seed(self.random_state + i + (n_estimators * workers_id))
```

Comment:
Тут же `i` это все таки не немой счетчик
```suggestion
            np.random.seed(self.random_state + tree_id + (n_estimators * workers_id))
```